feature/#81 예약 등록 검증 로직 변경

SD-gif opened the pull request:
## ⚡️ 관련 이슈
> 관련 있는 이슈를 태그해주세요.
- #81

## 📍주요 변경 사항
> 주요 변경 사항에 대해 작성해주세요.
- 예약 검증 로직 변경
- 로직 수행 위치 Command -> Facade 로 이동
- 코드 리팩터링
- 테스트 코드 리팩터링

## 🗣To Reviewer
> 의논이나 고려해야하는 내용을 작성해 주세요.

## 🌁 Screenshot
> (없을 경우 삭제) 작업한 내용에 대한 스크린샷을 첨부해주세요.

SD-gif committed:
feat/#81 예약 검증 로직 변경

SD-gif committed:
test/#81 검증 로직 수정에 따른 테스트 코드 수정

minjo-on committed:
refactor/#80 reviews 삭제 (#83)

fbehddn committed:
refactor/#84 강의 조회 응답 객체 분리 (#85)

SD-gif committed:
refactor: 예약 기능 리팩터링

SD-gif committed:
refactor: import 정리

SD-gif committed:
test: 테스트 코드 리팩터링

minjo-on commented on saerojinro-api/src/main/java/goorm/saerojinro/api/reservation/application/ReservationFacade.java:
```diff
@@ -26,20 +26,21 @@ public ReservationCreateResponse create(Long userId, Long lectureId) {
         User user = getUser(userId);
         Lecture lecture = getLecture(lectureId);
 
```
현재 방식에서는 **객체를 조회한 후 다시 조회 쿼리를 실행**하기 때문에, 불필요한 조회가 **총 3번** 발생합니다. 하지만 JPA에서는 **객체 없이 ID만으로도 조회가 가능**하므로, 처음부터 ID를 매개변수로 전달하면 **한 번의 쿼리로 끝낼 수 있습니다.**  

🔹 **기존 방식 (조회 3번 발생)**  
1. `User` 조회  
2. `Lecture` 조회  
3. `Reservation` 조회  

🔹 **개선 방식 (조회 1번으로 최적화)**  
```java
boolean existsByUserIdAndLectureId(Long userId, Long lectureId);
```
- **JPA는 ID만으로도 동일한 조회가 가능**하므로, 불필요한 객체 로딩을 제거  
- **조회 쿼리 횟수를 1번으로 줄여 성능 최적화**

SD-gif replied:
수정 하겠습니다!

SD-gif replied:
현재 API 명세서 상 강의 생성 요청에는 유저 아이디, 강의 아이디가 포함되어 있습니다.

생성 로직 검증 과정에서 강의 시작 시간이 필요합니다.
이 때, 클라이언트에서 강의 시간을 받아오기 보단, 데이터베이스에서 강의 시간을 가져오는 것이 바람직하다고 생각합니다.
하지만, 이렇게 되면 강의 조회는 불가피 하게 됩니다.

1. 강의 조회
2. 예약 조회(검증)
3. 예약 생성

이런식으로 진행 하려고합니다. 의견 부탁드립니다!

minjo-on replied:
Lazy 로딩을 위한 FetchType.LAZY 설정을 했고, 트랜잭션을 붙였기 때문에 Hibernate 프록시가 알아서 reservation.lecture.startedAt 이런식으로 강의에 대한 정보나 유저에 대한 정보를 가져올 수 있어요

minjo-on replied:
각 Id를 통해 Reservation 객체를 가져왔고, facade 트랜잭션 메소드 내에서는 reservation에 있는 User와 Lecture에 대한 정보를 가져올 수 있습니다.

minjo-on commented on saerojinro-api/src/main/java/goorm/saerojinro/api/reservation/application/ReservationFacade.java:
```diff
@@ -26,20 +26,21 @@ public ReservationCreateResponse create(Long userId, Long lectureId) {
-        Reservation reservation = reservationCommandService.create(user, lecture);
+        if(reservationQueryService.existsCheckByStartTime(user, lecture)){
+            throw new ReservationExistException();
+        }
```
Facade는 여러 서비스를 조합하는 역할을 하는 레이어이므로, 예외 처리 로직은 QueryService에서 수행하는 것이 더 적절합니다. 데이터 조회와 관련된 예외(EntityNotFoundException 등)를 QueryService에서 처리하면, Facade는 흐름을 관리하는 역할에 집중할 수 있어 책임이 명확해지고 유지보수가 쉬워집니다.

SD-gif replied:
중복을 방지하지 하기 위해 데이터 베이스에서 존재여부를 판단하는 로직은 생성할 때 필요하다고 생각 했습니다. 

커맨드 서비스 내에 create 메소드 에서 reservationRepository.exists 을 사용해 로직 검증 하도록 수정할까요?

minjo-on replied:
Queryservice에서 검증한 뒤, Command에서 저장하면 될 것 같습니다

minjo-on commented on saerojinro-api/src/main/java/goorm/saerojinro/api/reservation/application/ReservationFacade.java:
```diff
@@ -26,20 +26,21 @@ public ReservationCreateResponse create(Long userId, Long lectureId) {
+    public void cancel(Long userId, Long lectureId){
         User user = getUser(userId);
         Lecture lecture = getLecture(lectureId);
 
```
이 부분도 위에서 말씀드린대로 변경하면 될 것 같아요

SD-gif replied:
반영하겠습니다.

LeeShinHaeng reviewed (approved):
Facade에서 체크하며 CommandService 측의 QueryService 주입이 빠진 부분이 인상적이네요 👍
LGTM!

minjo-on commented on saerojinro-api/src/main/java/goorm/saerojinro/api/reservation/application/ReservationFacade.java:
```diff
@@ -26,20 +26,21 @@ public ReservationCreateResponse create(Long userId, Long lectureId) {
         Lecture lecture = getLecture(lectureId);
 
-        Reservation reservation = reservationCommandService.create(user, lecture);
+        if(reservationQueryService.existsCheckByStartTime(user, lecture)){
```
현재 로직이 예약이 중복인지와 강의가 시작 전인지 확인하는 로직인거 같은데 중복 예약 여부와 강의 시작 여부는 각각 Reservation과 Lecture 도메인의 책임이며, 하나의 쿼리로 합칠 경우 불필요한 JOIN이 발생하고, 특정 로직만 재사용하기 어려워 유지보수가 복잡해지므로, 책임을 분리하고 성능을 최적화하기 위해 두 개의 별도 쿼리로 처리하는 것이 더 적절해보입니다 !

SD-gif replied:
현재 로직은 유저가 가진 예약 내에서 예약하려는 강의의 시작 시간과 동일한 시작 시간을 가진 예약(다른 강의)이 있을 경우 예외를 반환 하도록 구상 하였습니다. 즐겨찾기 테이블에 있는 강의 아이디를 통해 다시 강의 테이블을 조회 해야 해서 성능이 우려되긴 하는데, 어떤 방식이 좋을지 의견 부탁드립니다

minjo-on replied:
그러면 쿼리를 existsByUserIdAndStartTime 이렇게 하고 true가 나오면 에러를 던지도록 하는건 어떨까요? Query annotation을 사용하지 않고 간단하게 누구나 알아볼 수 있을 것 같아요

minjo-on replied:
```java
@Query("SELECT COUNT(r) > 0 FROM Reservation r " +
       "WHERE r.user.id = :userId " +
       "AND r.lecture.startTime = :startTime")
boolean existsByUserAndStartTime(@Param("userId") Long userId, 
                                 @Param("startTime") LocalDateTime startTime);
``` 
 이렇게 하면 조인을 하지 않고 쿼리를 작성할 수 있을 것 같아요 추후에 redis를 이용해서 캐싱 적용하면 좋을 것 같습니다 !

SD-gif replied:
조인 없이도 이런 방식으로 접근할 수 있군요! 감사합니다 😊
반영하겠습니다 :)

minjo-on reviewed (commented):
고생하셨습니다 ! 코멘트 확인 부탁드려요

minjo-on committed:
refactor/#88 admin 모듈 정리 (#89)

LeeShinHaeng committed:
feature/#79 시간표 조회 (#86)

SD-gif committed:
refactor: Reservation 생성 메소드 명 변경에 따른 리팩터링

SD-gif committed:
refactor: Reservation 리팩터링

SD-gif committed:
refactor: Reservation 리팩터링

# Conflicts:
#	saerojinro-domain/src/main/java/goorm/saerojinro/domain/reservation/application/ReservationQueryService.java

SD-gif committed:
test: 테스트 코드 리팩터링

SD-gif committed:
refactor: Reservation 코드 리팩터링

SD-gif committed:
refactor: 미사용 import 제거

SD-gif committed:
refactor: 충돌제거

SD-gif committed:
Merge branch 'dev' into feature/#81-예약-등록-검증-로직-변경

SD-gif committed:
refactor: 매게변수 변경

SD-gif committed:
Merge remote-tracking branch 'origin/feature/#81-예약-등록-검증-로직-변경' into feature/#81-예약-등록-검증-로직-변경

SD-gif committed:
test: 테스트 코드 추가
